docs/source/extending_pywr/extending_pywr_parameters.rst

```diff
@@ -0,0 +1,224 @@
+
+.. code-block:: yaml
+
+    "max_flow": {
+        "type": "myparameter",
+        "value": 123.0
+    }
+
+Timesteps and scenarios
+=======================
+
+The ``Parameter.value`` method is called once per timestep for each scenario. The value it returns can be varied using
+the ``timestep`` and ``scenario_index`` arguments. For example, a simple version of a ``MonthlyProfileParameter`` could be
+created using the following:
+
+.. code-block:: python
+
+    class MonthlyProfileParameter(Parameter):
+        def __init__(self, model, profile):
+            super().__init__(model)
+            self.profile = profile  # a 12-element list of floats
+
+        def value(self, timestep, scenario_index):
+            index = timestep.month - 1  # convert to zero-based index
+            value = self.profile[index]
+
+        @dataclass
+        def load(cls, model, data):
+            profile = data.pop(profile)
+            return cls(model, profile)
+
+Tracking state with setup, reset, before and after
+==================================================
+
+The ``Parameter.setup`` and ``Parameter.reset`` methods are called once at the start of a model run before the first
+timestep. The ``reset`` method is called for every run, while the `setup` method is only called if the structure of
+the model has changed [*]_.
+
+The ``Parameter.before`` and ``Parameter.after`` methods are called before and after each timestep, respectively. These
+methods can be used when a parameter needs to track state between timesteps. For example, a licence parameter needs
+to track the volume remaining in the licence. It's important to remember that when using scenarios the model has
+multiple states. It's good practice to write stateful parameters with this in mind, even if you aren't using scenarios
+initially, so that you can in the future without rewriting anything. The example below shows a very simplistic licence
+parameter which has a finite volume.
+
+.. code-block:: python
+
+    class LicenceParameter(Parameter):
+        def __init__(self, model, total_volume):
+            super().__init__(self, model)
+            self.total_volume = total_volume
+
+        def setup(self):
+            # allocate an array to hold the parameter state
+            num_scenarios = len(self.model.scenarios.combinations)
+            self._volume_remaining = np.empty([num_scenarios], np.float64)
+
+        def reset(self):
+            # reset the amount remaining in all states to the initial value
+            self._volume_remaining[...] = self.total_volume
+
+        def value(self, timestep, scenario_index):
+            # return the current volume remaining for the scenario
+            return self._volume_remaining[scenario_index.global_id]
+
+        def after(self):
+            # update the state
+            timestep = self.model.timestepper.current  # get current timestep
+            flow_during_timestep = self._node.flow * timestep.days  # array of flows
```

Comment:
`self._node` is not explained here. This only works if this parameter is assigned directly to a node?